miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/RunRestController.java

```diff
@@ -174,4 +248,55 @@ public void setSecurityManager(SecurityManager securityManager) {
     return RestExceptionHandler.handleException(request, response, exception);
   }
 
+  @RequestMapping(value = "{runId}/add", method = RequestMethod.POST, produces = "application/json")
+  @ResponseStatus(code = HttpStatus.OK)
+  public void addContainerByBarcode(@PathVariable Long runId, @RequestParam("barcode") String barcode) throws IOException {
+    Run run = runService.get(runId);
+    Collection<SequencerPartitionContainer> containers = containerService
+        .listByBarcode(barcode);
+    if (containers.isEmpty()) {
+      throw new RestException("No containers with this barcode.", Status.BAD_REQUEST);
+    }
+    if (containers.size() > 1) {
+      throw new RestException("Multiple containers with this barcode.", Status.BAD_REQUEST);
+    }
+    SequencerPartitionContainer container = containers.iterator().next();
+    if (container.getPlatform().getId() != run.getSequencerReference().getPlatform().getId()) {
+      throw new RestException(String.format("This container is meant for %s but this run came from %s.",
+          container.getPlatform().getNameAndModel(), run.getSequencerReference().getPlatform().getNameAndModel()), Status.BAD_REQUEST);
+    }
+    run.addSequencerPartitionContainer(container);
+    runService.update(run);
+  }
+
+  @RequestMapping(value = "{runId}/remove", method = RequestMethod.POST, produces = "application/json")
+  @ResponseStatus(code = HttpStatus.OK)
+  public void removeContainer(@PathVariable Long runId, @RequestBody List<Long> containerIds) throws IOException {
+    Run run = runService.get(runId);
+    run.setSequencerPartitionContainers(run.getSequencerPartitionContainers().stream()
+        .filter(container -> !containerIds.contains(container.getId())).collect(Collectors.toList()));
+    runService.update(run);
+  }
+
+  @RequestMapping(value = "{runId}/qc", method = RequestMethod.POST, produces = "application/json")
+  @ResponseStatus(code = HttpStatus.OK)
+  public void setQc(@PathVariable Long runId, @RequestBody RunPartitionQCRequest request) throws IOException {
+    Run run = runService.get(runId);
+    PartitionQCType qcType = partitionQCService.listTypes().stream().filter(qt -> qt.getId() == request.getQcTypeId()).findAny().get();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "Optional#isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3655)